velox/serializers/RowSerializer.h

```diff
@@ -145,25 +277,59 @@ class RowDeserializer {
   static void deserialize(
       ByteInputStream* source,
       std::vector<SerializeView>& serializedRows,
-      std::vector<std::unique_ptr<std::string>>& serializedBuffers) {
+      std::vector<std::unique_ptr<std::string>>& serializedBuffers,
+      const VectorSerde::Options* options) {
+    const auto compressionKind = options == nullptr
+        ? VectorSerde::Options().compressionKind
+        : options->compressionKind;
     while (!source->atEnd()) {
-      // First read row size in big endian order.
-      const auto rowSize = folly::Endian::big(source->read<TRowSize>());
-      auto serializedBuffer = std::make_unique<std::string>();
-      serializedBuffer->reserve(rowSize);
-
-      const auto row = source->nextView(rowSize);
-      serializedBuffer->append(row.data(), row.size());
-      // If we couldn't read the entire row at once, we need to concatenate it
-      // in a different buffer.
-      if (serializedBuffer->size() < rowSize) {
-        concatenatePartialRow(source, rowSize, *serializedBuffer);
+      std::unique_ptr<folly::IOBuf> uncompressedBuf = nullptr;
+      const auto header = detail::RowHeader::read(source);
+      if (header.compressed) {
+        VELOX_DCHECK(
+            compressionKind != common::CompressionKind::CompressionKind_NONE);
+        auto compressBuf = folly::IOBuf::create(header.compressedSize);
+        source->readBytes(compressBuf->writableData(), header.compressedSize);
+        compressBuf->append(header.compressedSize);
+
+        // Process chained uncompressed results IOBufs.
+        const auto codec = common::compressionKindToCodec(compressionKind);
+        uncompressedBuf =
+            codec->uncompress(compressBuf.get(), header.uncompressedSize);
+      }
+      std::unique_ptr<ByteInputStream> uncompressedStream;
+      ByteInputStream* uncompressedSource;
+      if (uncompressedBuf == nullptr) {
+        uncompressedSource = source;
+      } else {
+        uncompressedStream = std::make_unique<BufferInputStream>(
+            byteRangesFromIOBuf(uncompressedBuf.get()));
+        uncompressedSource = uncompressedStream.get();
       }
+      const std::streampos initialSize = uncompressedSource->tellp();
+      while (uncompressedSource->tellp() - initialSize <
+             header.uncompressedSize) {
+        // First read row size in big endian order.
+        const auto rowSize =
+            folly::Endian::big(uncompressedSource->read<TRowSize>());
```

Comment:
If we read in big endian, then shall we also write in big endian?

Reply:
Yes, the logic is same as before, just move it.